perceval/utils/density_matrix.py

```diff
@@ -0,0 +1,361 @@
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import scipy.sparse.linalg
+
+from perceval.utils.statevector import StateVector, SVDistribution, BasicState, max_photon_state_iterator
+from typing import Union, Optional
+from math import comb
+from numpy import conj
+import numpy as np
+from scipy.sparse import dok_array, sparray, csr_array, kron
+from scipy.sparse.linalg import LinearOperator, eigsh
+from scipy.linalg import eigh
+from copy import copy
+import random
+
+
+def create_index(m, n_max):
+    index = dict()
+    for i, x in enumerate(max_photon_state_iterator(m, n_max)):
+        index[x] = i
+    return index
+
+
+def statevector_to_array(sv, index):
+    """
+    translate a StateVector object into an array
+    :param sv: a StateVector
+    :param index: a dictionary with BasicStates as keys and indices as values
+    """
+    vector = np.zeros(len(index),  dtype=complex)
+    for key, value in sv:
+        vector[index[key]] += value
+    return vector
+
+
+def array_to_statevector(vector, reverse_index):
+    """
+    translate an array in a StateVector
+    :param vector: an array
+    :param reverse_index: a list of BasicStates
+    """
+
+    sv = 0*reverse_index[0]
+    for i, x in enumerate(reverse_index):
+        if vector[i] == 0:
+            continue
+        else:
+            sv += complex(vector[i])*x
+    return sv
+
+
+def density_matrix_tensor_product(A, B):
+    """
+    Make the tensor product of 2 Density Matrices
+    :param A, B: two density matrices
+    :return: the "kronecker" product of the density matrices, in the correct basis
+    """
+
+    if isinstance(A, (SVDistribution, StateVector, BasicState)):
+        A = DensityMatrix(A)
+
+    if isinstance(B, (StateVector, SVDistribution, BasicState)):
+        B = DensityMatrix(B)
+
+    if not isinstance(B, DensityMatrix):
+        raise TypeError(f"Cannot do a Tensor product between a DensityMatrix and a {type(B)}")
+
+    if not isinstance(A, DensityMatrix):
+        raise TypeError(f"Cannot do a tensor product between a {type(A)} and a DensityMatrix")
+
+    n_max = A.n_max + B.n_max
+    n_mode = A.m + B.m
+    size = comb(n_max+n_mode, n_max)
+    new_index = create_index(n_mode, n_max)
+    matrix = kron(A.mat, B.mat)
+    perm = dok_array((A.size*B.size, size), dtype=complex) # matrix from tensor space to complete Fock space
+    for i, a_state in enumerate(A.reverse_index):
+        for j, b_state in enumerate(B.reverse_index):
+            index = new_index[a_state*b_state]
+            perm[i*B.size+j, index] = 1
+
+    matrix = perm.T @ matrix @ perm
+
+    return DensityMatrix(matrix, new_index)
+
+
+class DensityMatrix:
```

Comment:
General comment about this class: could you add typing hints in the method signatures?